undertow-servlet/src/test/java/org/wildfly/elytron/web/undertow/server/servlet/FormServletAuthenticationWithClusteredSSOTest.java

```diff
@@ -0,0 +1,255 @@
+
+        assertLoginPage(httpClient.execute(new HttpGet(serverA.createUri())));
+
+        assertFalse(cookieStore.getCookies().stream().filter(cookie -> cookie.getName().equals("JSESSIONSSOID")).findAny().isPresent());
+
+        // log into APP_A
+        HttpResponse execute = loginToApp(httpClient, serverA, "ladybird", "Coleoptera");
+        assertTrue(cookieStore.getCookies().stream().filter(cookie -> cookie.getName().equals("JSESSIONSSOID")).findAny().isPresent());
+        assertSuccessfulResponse(execute, "ladybird");
+        String appOneSessionId = getSessionIdForApp(cookieStore, serverA);
+
+        // can now access APP_B without logging in again
+        assertSuccessfulResponse(httpClient.execute(new HttpGet(serverB.createUri())), "ladybird");
+        String appTwoSessionId = getSessionIdForApp(cookieStore, serverB);
+
+        // log out of APP_A
+        httpClient.execute(new HttpGet(serverA.createUri("/logout")));
+
+        // log into APP_A again
+        execute = loginToApp(httpClient, serverA, "ladybird", "Coleoptera");
+        assertTrue(cookieStore.getCookies().stream().filter(cookie -> cookie.getName().equals("JSESSIONSSOID")).findAny().isPresent());
+        assertSuccessfulResponse(execute, "ladybird");
+        String appOneNewSessionId = getSessionIdForApp(cookieStore, serverA);
+
+        // the session ID for APP_A should now be different from the initial session ID
+        assertTrue(appOneSessionId != null && appOneNewSessionId != null && ! appOneSessionId.equals(appOneNewSessionId));
+
+        // access APP_B without logging in again
+        assertSuccessfulResponse(httpClient.execute(new HttpGet(serverB.createUri())), "ladybird");
+        String appTwoNewSessionId = getSessionIdForApp(cookieStore, serverB);
+
+        // the session ID for APP_B should now be different from the initial session ID
+        assertTrue(appTwoSessionId != null && appTwoNewSessionId != null && ! appTwoSessionId.equals(appTwoNewSessionId));
+    }
+
+    private static HttpResponse loginToApp(HttpClient httpClient, UndertowServer server, String username, String password) throws Exception {
+        assertLoginPage(httpClient.execute(new HttpGet(server.createUri())));
+        HttpPost httpAuthenticate = new HttpPost(server.createUri("/j_security_check"));
+        List<NameValuePair> parameters = new ArrayList<>(2);
+        parameters.add(new BasicNameValuePair("j_username", "ladybird"));
+        parameters.add(new BasicNameValuePair("j_password", "Coleoptera"));
+        httpAuthenticate.setEntity(new UrlEncodedFormEntity(parameters));
+        return httpClient.execute(httpAuthenticate);
+    }
+
+    private static String getSessionIdForApp(BasicCookieStore cookieStore, UndertowServer server) {
+        return cookieStore.getCookies().stream().filter(cookie -> cookie.getName().equals("JSESSIONID") && cookie.getPath().equals(server.getContextRoot())).findAny().get().getValue();
```

Comment:
It is not part of our requirements at the moment, but I am finding as screens get wider restricting code to a reasonable line length becomes more relevant again as now it is possible to open multiple columns of source on a single screen.